### ISSUE TITLE
Sending messages via SQS with X-Ray Tracing headers

### ISSUE THREAD

**Issue by bml1g12** [awslabs/aws-lambda-powertools-python#659]
**Is your feature request related to a problem? Please describe.**
I have tested the following lambda function

```
import logging
LOGGER = Logger(datefmt="%Y-%m-%dT%H:%M:%SZ")
METRICS = Metrics()
# Tracer automatically patches all supported libraries by X-Ray during initialization, by default.
# https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-python-patching.html
TRACER = Tracer()  # Sets service via env var
initialise_sentry()

@TRACER.capture_lambda_handler(capture_error=True)
@METRICS.log_metrics(capture_cold_start_metric=False)
def hello(event, context):
    """This function will loop infinitely until the lambda timeout occurs and kills it"""
    LOGGER.info("Starting hello lambda")
    region = "us-east-1"
    sqs_client = boto3.client("sqs",
                              region_name=region,
                              config=botocore.client.Config(max_pool_connections=25))
    response = sqs_client.get_queue_url(
        QueueName="deletme-sqs",
    )
    print("sending message")
    sqs_client.send_message(
        QueueUrl=response["QueueUrl"],
        MessageBody=json.dumps("test message"))
    print("message sent")
```

```
╰─❯ aws sqs receive-message --region us-east-1 --queue-url my-queue-url--attribute-names All --message-attribute-names All --max-number-of-messages 10                                  ─╯
{
    "Messages": [
        {
            "MessageId": "6a5ee039-575c-40ae-bbfd-b67a54ce6aa0",
            "ReceiptHandle": "AQEBpD+AFjum1UsJo4UtVvzGWI8uItsr5gf5PgKeUr69JMgxse3Zk8pUk58xcuBSH1Lns6uIXlZqwOwyPAw0nZfifPmrhWuu4oqnoISP7ondNGx9mFy05SiICZIK+RjyYDvy9wU+mvNTrxL2Wz71Mdtz094Xkq9AURMgZsTP3D+8odRgj5g+Qssma4mQJ9neM4chiZ0DKIWDA/p32BOVUfiatIQz+KzCpGBnw9dmA1FlgNdc2sitw4AeT0Z0Z74vsIThKNNCendYWDrJT5lknDend8TlMTaAEeWKFyWKuwO/0nIT+41Gmii7jUeYPAV3t9CuMZqm/y+mwjb0hSoMuL2pZDhvt8jgkyob4wNQ5UxG0lP/3PiBeAPzdZhclZ3mK9T7",
            "MD5OfBody": "224adf45124f9ba44f83f49cc7964687",
            "Body": "\"test message\"",
            "Attributes": {
                "SenderId": "AROAX3RAZYFVIC5IA6PER:errorhandlingdemo-development-helloworld",
                "ApproximateFirstReceiveTimestamp": "1630087501286",
                "ApproximateReceiveCount": "1",
                "SentTimestamp": "1630087499030"
            }
        }
    ]
}
````

**Describe the solution you'd like**
I was expecting the output to message header to include `X-Amzn-Trace-Id` because here 

 https://docs.aws.amazon.com/xray/latest/devguide/xray-services-sqs.html 
> Default HTTP Header – The X-Ray SDK automatically populates the trace header as an HTTP header when you call Amazon SQS through the AWS SDK. The default trace header is carried by X-Amzn-Trace-Id and corresponds to all messages included in a SendMessage or SendMessageBatch request. To learn more about the default HTTP header, see Tracing header.

And I thought that the Tracing() powertool class enables botocore X-Ray SDK instrumentation. 

**Describe alternatives you've considered**
I assume I could manually add this attribute to the send_message() boto call, but I thought this would be handled by the wrapper automagically. 

I also tried @TRACER.capture_method but got the same result.

**Comment by walmsles** [comment]
Including the AWS Xray SDK means when you activate X-Ray tracing on the Lambda at the infrastructure level it will send data to the X-Ray Service.

Is XRay tracing enabled on your lambda?

**Comment by bml1g12** [comment]
I'm using the serverless framework to enable it, but I am fairly sure it is enabled as I see the X-Ray traceId is logged 

It has 
```
Allow: xray:PutTraceSegment
Allow: xray:PutTraceSegments
Allow: xray:PutTelemetryRecords
```
For IAM, and the GUI shows the following

![image](https://user-images.githubusercontent.com/2605767/131228673-822c1e3d-2e64-40b7-a69d-d22296da22fb.png)

**Comment by bml1g12** [comment]
Would be grateful for any thoughts on this one

**Comment by walmsles** [comment]
Apologies for the lack of response.  I have put together a small reference project available [here](https://github.com/walmsles/aws-serverless-api-sqs) which has API Gateway --> Lambda --> SQS --> Lambda.  This example when I deploy has the header value for X-Ray passed through.

The passing of the header value through the SQS interaction is a feature of the SDK itself and not a feature of the Python Powertools library (other than the patching).

My example does not patch explicitly with Tracer as in your example so ALL modules will be patched which works.  I wonder if you need to also add 'requests' to your patch list since is likely used by boto3 (also in the example provided in powertools documentation).  The patching behaviour will only patch the modules you pass in ignoring all others.

Let me know how you go  - will have time later to try changing my project to match yours in a few hours.

**Comment by walmsles** [comment]
Updated my repo to include patching just "boto3" and still get the same behaviour.

Here is the Event from SQS - note the "AWSTraceHeader" in the "attributes" - this is with X-Ray tracing turned OFF in the console and via the serverless framework.

@bml1g12 are you able to share an example that is not working for you?  

```python
{
    "level": "INFO",
    "location": "process_it:18",
    "message": {
        "messageId": "7c4dc93c-684b-440f-8106-2382b688ef6a",
        "receiptHandle": "AQEB2Rc/FrOl2TuY+wkBRjZ2d42bmF0Tb0u1nrcKTZvjobp62QEJ7yMvbSQ6z8B1Eds6Ku4okF6C7DXyhHVqEdc7MEWI5+s1e55L+Dd4G8Ajk8oxB+mS22We7dV+pI98RxlAYPg/vdV1aY7yw7uXY538+JpPkcvj1u1LaIXHeCraLwU53ufktZPVHjnUvrw7LNZQ6Jmvw8MgO0Z7bE16JOMcgW88PFKjNJJ4hGM7JEQ8keByCd3DdIYR3k1NkWcCYdHDlyc2AqO7dYRYKGpymJj22cPgRoqsDozjA0xrBpd1yB/P46q0M7vVBpObZ0im8BltXMUvwsQ81+7vwE/rERUBcLOU6idWH6eF1qbgNsbyrhF/NNZxC5bTRpWXRSTWttu6DsuwFvQia71DwmFVQPLUoO2/AMLkkP5coxUHDoYL9rsqEjY39P0yk7o668LCSlI8",
        "body": "{\n    \"message\": \"Test message\"\n}",
        "attributes": {
            "ApproximateReceiveCount": "8",
            "AWSTraceHeader": "Root=1-61386449-152d91c220630c6b764dbcfc;Parent=0000000000000000;Sampled=0",
            "SentTimestamp": "1631085642313",
            "SenderId": "AROAZZ4DN6W6GIUBAIUEX:api-sqs-dynamodb-dev-api",
            "ApproximateFirstReceiveTimestamp": "1631085642313"
        },
        "messageAttributes": {},
        "md5OfBody": "480beaa3c727a6a5181acc4047cd4e01",
        "eventSource": "aws:sqs",
        "eventSourceARN": "arn:aws:sqs:ap-southeast-2:xxx_account_id_xxx:api-sqs-dynamodb-dev-sqs-12Z1CRWZNBI83",
        "awsRegion": "ap-southeast-2"
    },
    "timestamp": "2021-09-08 07:24:14,999+0000",
    "service": "service_undefined",
    "cold_start": false,
    "function_name": "api-sqs-dynamodb-dev-sqs-processor",
    "function_memory_size": "1024",
    "function_arn": "arn:aws:lambda:ap-southeast-2:xxx_account_id_xxx:function:api-sqs-dynamodb-dev-sqs-processor",
    "function_request_id": "af7d4d05-b246-52e4-af11-14b7d50f4ce3",
    "xray_trace_id": "1-6138651e-0192b46a2b08ff1c45893f4c"
}
```

**Comment by bml1g12** [comment]
I see so you can't reproduce the issue I am having. I'll set up a repo to reproduce the issue exactly.

**Comment by bml1g12** [comment]
That's odd, I made a [minimal example repo](https://github.com/bml1g12/PowertoolsTracingExample), and it works fine (AWSTraceHeader appears) suggesting that as you say, that the normal behaviour of the library works as expected. Which is great...!

But I can use serverless to remove a stack and re-deploy a fresh stack reproducibly which has the issue. This stack that exhibits the issue has a more complex serverless.yml with extra bells and whistles that I would have thought would be irrelevant to this, but has exactly the same handler function as the minimal example that works fine. I can't share this without working on it a bit as it might contain some confidential information.

As this looks like a time consuming one to debug, I think what I'll do is put this on my backlog of work to look into in more detail and report back with minimal example to reproduce the issue. Given the handler is the same between working and non-working cases, it seems likely the issue is related to:

* Serverless framework deployment causing some difference in the python versions of things being deployed
* The SQS queue permissions tied to the queue itself  

So probably not related to aws-lambda-powertools directly; although probably relevant to the userbase of this repo.

**Comment by walmsles** [comment]
Thanks so much for your efforts.  Perhaps @marekq  or @heitorlessa may have some insights?

I am still curious about the outcome tbh, the only way I could get this to happen was to remove the Tracer altogether.

Are you packaging python dependencies with serverless-python-requirements?

My only theory is conflicting library versions - perhaps compare boto3 versions between the projects or remove any dependencies cached by serverless-python-requirements (if in use).

**Comment by bml1g12** [comment]
I created a separate repo with identical code as the failing lambda but a different serverless "service name", and that one worked fine.

I downloaded the lambda deployment folder for the not working lambda and the working lambda and everything seems the same, so I am completely stumped by what causes one deployment to have this issue. I am going to write it off an obscure edge case that cannot be reproduced in a fresh deployment. 

Thanks for your help in identifying that this is not a bug in the PowerTools codebase.
